app/src/main/java/org/mozilla/fenix/wallpapers/WallpapersUseCases.kt

```diff
@@ -0,0 +1,246 @@
+        val downloader = WallpaperDownloader(context, client)
+        DefaultInitializeWallpaperUseCase(
+            store = store,
+            downloader = downloader,
+            fileManager = fileManager,
+            settings = context.settings(),
+            currentLocale = currentLocale
+        )
+    }
+    val loadBitmap: LoadBitmapUseCase by lazy { DefaultLoadBitmapUseCase(context) }
+    val selectUseCase: SelectWallpaperUseCase by lazy { DefaultSelectWallpaperUseCase(context.settings(), store) }
+
+    /**
+     * Contract for usecases that initialize the wallpaper feature.
+     */
+    interface InitializeWallpapersUseCase {
+        /**
+         * Start operations that should be down during initialization, like remote metadata
+         * retrieval and determining the currently selected wallpaper.
+         */
+        suspend operator fun invoke()
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal class DefaultInitializeWallpaperUseCase(
+        private val store: AppStore,
+        private val downloader: WallpaperDownloader,
+        private val fileManager: WallpaperFileManager,
+        private val settings: Settings,
+        private val currentLocale: String,
+        private val possibleWallpapers: List<Wallpaper> = allWallpapers,
+    ) : InitializeWallpapersUseCase {
+
+        /**
+         * Downloads the currently available wallpaper metadata from a remote source.
+         * Updates the [store] with that metadata and with the selected wallpaper found in storage.
+         * Removes any unused promotional or time-limited assets from local storage.
+         */
+        override suspend operator fun invoke() {
+            // Quite a bit of code needs to be executed off the main thread in some of this setup.
+            // This should be cleaned up as improvements are made to the storage, file management,
+            // and download utilities.
+            withContext(Dispatchers.IO) {
+                val availableWallpapers = getAvailableWallpapers()
+                val currentWallpaperName = settings.currentWallpaper
+                val currentWallpaper = possibleWallpapers.find { it.name == currentWallpaperName }
+                    ?: fileManager.lookupExpiredWallpaper(currentWallpaperName)
+                    ?: Wallpaper.Default
+
+                fileManager.clean(
+                    currentWallpaper,
+                    possibleWallpapers.filterIsInstance<Wallpaper.Remote>()
+                )
+                downloadAllRemoteWallpapers(availableWallpapers)
+                store.dispatch(AppAction.WallpaperAction.UpdateAvailableWallpapers(availableWallpapers))
+                store.dispatch(AppAction.WallpaperAction.UpdateCurrentWallpaper(currentWallpaper))
+            }
+        }
+
+        private fun getAvailableWallpapers() = possibleWallpapers
+            .filter(::filterExpiredRemoteWallpapers)
+            .filter(::filterPromotionalWallpapers)
```

Comment:
Would this potentially be more efficient?
```suggestion
            .filter { filterExpiredRemoteWallpapers(it) && filterPromotionalWallpapers(it) } 
```

Reply:
Oh yeah, that definitely seems better. IIRC we added them separately which is probably led to that. I don't really like the names much either TBH. Maybe something like `.filter { !it.isExpired() && !it.isUnavailablePromotion() }` would convey the meaning a bit better?

Reply:
oooo, yeah, making extension or class functions and cleaning up the syntax would make it look crisp.

Reply:
Updated!